File: src/Parser.cpp
#include "Parser.hpp"

Parser::Parser(Lexer lexer)
{
	*lex = lexer;

	*current_token = lex->next_token();
}

void Parser::parse()
{
	function_list();
}

void Parser::next_token()
{
	*current_token = lex->next_token();
}

bool Parser::check_next(Token_Type type)
{
	if (!test_check_next(type))
	{
		unexpected_token();	
		return false;
	}

	return true;
}

bool Parser::check_end()
{
	return (test_check_next(TK_EOF) || test_check_next(TK_EOL) || test_check_next(TK_SEMICOLON));
}

bool Parser::test_check_next(Token_Type type)
{
	if (current_token->type == type)
	{
		next_token();
		return true;
	}

	return false;
}

void Parser::unexpected_token()
{
	std::cout << "unexpected token " + lex->type2name(current_token->type) << std::endl;
	exit(0);
}

void Parser::eat_eol()
{
	while (test_check_next(TK_EOL));
}

/*

				Grammar Rules

*/

bool Parser::block_follow()
{
	switch (current_token->type)
	{
		case TK_ELSIF:
		case TK_ELSE:
		case TK_END:
		case TK_EOF:
			return true;
		default:
			return false;
	}
}

struct precedence Parser::get_precedence(Token_Type type)
{
	for (unsigned int i = 0; i < priorities.size(); i++)
		if (type == priorities[i].operator_type)
			return priorities[i];

	return (struct precedence){0, true, 0};
}

ExprAST *Parser::subexpression(ExprAST *lhs, int limit)
{
	while (true)
	{	
		struct precedence expr_operator = get_precedence(current_token->type);

		if ((limit > expr_operator.precedence) || (expr_operator.precedence == 0))
			return lhs;

		Token_Type operator_type = current_token->type;
		next_token();

		ExprAST *rhs = simple_expression();

		struct precedence next_operator = get_precedence(current_token->type);

		if (next_operator.precedence == 0)
			return new BinaryExprAST(operator_type, lhs, rhs);

		if (!expr_operator.left) {
			if (next_operator.precedence >= expr_operator.precedence)
				rhs = subexpression(rhs, limit + 1);
		} else {
			if (next_operator.precedence > expr_operator.precedence)
				rhs = subexpression(rhs, limit + 1);
		}
	
		lhs = new BinaryExprAST(operator_type, lhs, rhs);
	}
}

ExprAST *Parser::expression()
{
	ExprAST *lhs = simple_expression();
	return subexpression(lhs);
}

ExprAST *Parser::suffixed_expression()
{
	// primary_expression {'.' | '::' | '(' expression_list ')'}
	return primary_expression();
}

ExprAST *Parser::simple_expression()
{
	// number | true | false | nil

	switch(current_token->type)
	{
		case TK_NUMBER:
		{
			IntegerAST *number = new IntegerAST(current_token->value.ival);
			next_token();
			return number;
		}
		case TK_TRUE:
			next_token();
			break;
		case TK_FALSE:
			next_token();
			break;
		case TK_NIL:
			next_token();
			break;
		case TK_IDENTIFIER:
			next_token();

			//Don't forget the assignment AST here
			break;
		case TK_OPENPAR:
			next_token();
			expression();
			check_next(TK_CLOSEPAR);
			break;
		default:
			unexpected_token();
			break;
	}

	return NULL;
}

ExprAST *Parser::primary_expression()
{
	// Identifier | '(' expression ')'

	switch(current_token->type)
	{
		case TK_OPENPAR:
		{
			next_token();
			ExprAST *expr_ast = expression();
			check_next(TK_CLOSEPAR);
			return expr_ast;
		}
		case TK_IDENTIFIER:
		{
			IdentifierAST *ident = new IdentifierAST(*current_token->value.sval);
			next_token();
			return ident;
		}
		default:
		{
			unexpected_token();
			return NULL;
		}
	}
}

ExprAST *Parser::expr_statement()
{
	simple_expression();
	return NULL;
}

ExprAST *Parser::arglist()
{
	while (test_check_next(TK_IDENTIFIER) || test_check_next(TK_CONSTANT))
	{
		if (test_check_next(TK_COMMA))
		{
			// Insira vários nadas aqui
		}
	} 

	return NULL;
}

ExprAST *Parser::function()
{
	next_token(); // eat def

	if (test_check_next(TK_IDENTIFIER) || test_check_next(TK_CONSTANT))
	{
		eat_eol();
		
		while (current_token->type == TK_OPENPAR)
		{
			next_token();

			arglist();
			check_next(TK_CLOSEPAR);

			check_next(TK_EQUAL);

			eat_eol();

			statement_block();

			eat_eol();
		}
	}

	return NULL;
}

ExprAST *Parser::statement_block()
{
	if (test_check_next(TK_OPENBRAC))
	{
		while(current_token->type != TK_CLOSEBRAC)
		{
			std::cout << "C blocks" << std::endl;

			statement();
			eat_eol();
		}

		check_next(TK_CLOSEBRAC);
	} else {
		while(current_token->type != TK_EOL)
		{
			std::cout << "No blocks sorry" << std::endl;

			statement();

			if (current_token->type != TK_SEMICOLON)
				break;
		}
	}

	return NULL;
}

ExprAST *Parser::statement()
{
	switch(current_token->type)
	{
		case TK_SEMICOLON: 
		case TK_EOL:
			next_token();

			if (current_token->type != TK_EOF)
				return statement();

			break;
		case TK_IF:
			break;
		case TK_WHILE:
			break;
		default:
			return expr_statement();
			break;
	}

	return NULL;
}

ExprAST *Parser::function_list()
{
	switch(current_token->type)
	{
		case TK_DEF:
			return function();
		case TK_EOL:
			next_token();

			if (current_token->type != TK_EOF)
				return function_list();

			break;
		default:
			unexpected_token();
			break;
	}


	return NULL;
}

BlockAST *Parser::block()
{	
	std::vector<ExprAST*> statement_list;

	while (!block_follow())
	{
		ExprAST *stat = statement();

		if (stat != NULL)
			statement_list.push_back(stat);
		
		if (!check_end())
			unexpected_token();

		/*if (current_token->type == TK_RETURN)
		{
			statement_list.push_back(statement());
			break;
		}*/
	}

	return new BlockAST(statement_list);
}

File: src/Parser.rb
require './Lexer.rb'
require './Codegen.rb'

class Parser
    attr_reader :current_token
    attr_accessor :lexer

    def initialize lexer
        @lexer = lexer
        next_token
    end

    def parse
        function_list
    end

    def next_token
        @current_token = @lexer.next_token
    end

    def unexpected_token
        raise "MANO LARGA DE SER BURRO, SEU CÓDIGO ESTÁ ERRADO " + current_token.line.to_s + " " + current_token.type.to_s
    end

    def test_check_next token_type, ignore_eol = true
        test_check_next(token_type) if (current_token == :TK_EOL) and (ignore_eol)

        current_token.type == token_type ? (next_token; true) : (false)
    end

    def check_next token_type, ignore_eol = true
        return (next_token; check_next(token_type)) if (current_token.type == :TK_EOL) and (ignore_eol)

        current_token.type == token_type ? (next_token; true) : (puts token_type; unexpected_token; false)
    end

    def multiple_check_next tokens_type, ignore_eol = true
        return (next_token; multiple_check_next(tokens_type)) if (current_token.type == :TK_EOL) and (ignore_eol)

        tokens_type.include?(current_token.type) ? (next_token; true) : (unexpected_token; false)
    end

    def multiple_test_check_next tokens_type, ignore_eol = true
        return (next_token; multiple_test_check_next(tokens_type)) if (current_token.type == :TK_EOL) and (ignore_eol)

        tokens_type.include?(current_token.type) ? (next_token; true) : (false)
    end

    def check_end
        (test_check_next :TK_EOF) or (test_check_next :TK_EOL) or (test_check_next :SEMICOLON)
    end

    def eat_eol
        next_token while current_token.type == :TK_EOL
    end

    #----------------------------------------------------------------------------------------#
    #----------------------------------------------------------------------------------------#
    #                                   Grammar Rules                                        #
    #----------------------------------------------------------------------------------------#
    #----------------------------------------------------------------------------------------#

    def function_list
        list = []

        while true
            case current_token.type
            when :TK_DEF
                list << function
            when :TK_EOL
                next_token; next
            when :TK_EOF
                return list
            else
               # unexpected_token
            end
        end
    end

    def arglist essential = false, args = []
        argument = current_token

        if essential == true
            args << argument if multiple_check_next [:TK_IDENTIFIER, :TK_NUMBER], false
        else
            args << argument if multiple_test_check_next [:TK_IDENTIFIER, :TK_NUMBER], false
        end

        if current_token.type == :TK_COMMA
            next_token
            arglist true, args
        end

        return args
    end

    def type_definition essential = false, type_list = []
        type = current_token

        if essential == true
            type_list << type if check_next :TK_INT, false
        else
            type_list << type if test_check_next :TK_INT, false
        end

        if current_token.type == :TK_MINUS
            next_token
            check_next :TK_GREATER

            type_definition true, type_list
        end

        return type_list
    end

    def function
        next_token

        function_name = current_token.value
        func = Function.new function_name

        if (test_check_next :TK_IDENTIFIER, false)
            check_next :TK_DCOLON

            func_header = FunctionHeader.new type_definition

            while check_next :TK_OPENPAR
                args = arglist

                check_next :TK_CLOSEPAR, false

                check_next :TK_EQUAL, false

                stat = statement_block

                func.definitions << FunctionDefinition.new(args, stat)
                func.header = func_header

                eat_eol

                break if current_token.type != :TK_OPENPAR
            end

            check_next :TK_END
        else
            unexpected_token
        end

        return func
    end

    def statement_block
        statement_list = []

        if test_check_next :TK_OPENBRAC
            while current_token.type != TK_CLOSEBRAC
                statement_list << statement
            end
        else
            while current_token.type != :TK_EOL
                statement_list << statement
                
                break if current_token.type != :TK_SEMICOLON

                next_token
            end
        end

        return statement_list
    end

    def call_args args_list = []
        args_list << statement

        if current_token.type == :TK_COMMA
            next_token
            call_args args_list
        end

        return args_list
    end

    def simple_expr
        case current_token.type
        when :TK_IDENTIFIER
            ident = current_token
            next_token

            if test_check_next :TK_OPENPAR
                func = FunctionCall.new(ident.value)
                
                if !test_check_next :TK_CLOSEPAR
                    func.args = call_args
                end

                check_next(:TK_CLOSEPAR)

                return func
            else
                return Identifier.new(ident.value)
            end
        when :TK_NUMBER
            number = Number.new(current_token.value)
            next_token
            return number
        end
    end

    Priorities = { :TK_PLUS => [1, :left], :TK_MINUS => [1, :left],  
                   :"*"  => [2, :left], :"/" => [2, :left],
                   :"**" => [3, :right] }

    def is_binop? token
        Priorities[token] || [0, nil]
    end

    def expression lhs, limit = 0
        while true
            operator = is_binop?(current_token.type)

            if (operator[0] < limit) or (operator[0] == 0) 
                return lhs
            end

            bin_operator = current_token.type 
            next_token 

            rhs = simple_expr 

            next_operator = is_binop?(current_token.type) 

            if next_operator[0] == 0 
                return BinaryExpression.new lhs, rhs, bin_operator
            end

            if operator[1] == :right
                if (next_operator[0] >= operator[0]) 
                    rhs = subexpr rhs, operator[0] 
                end
            else
                if (next_operator[0] > operator[0])
                    rhs = subexpr rhs, operator[0] 
                end
            end

            lhs = BinaryExpression.new lhs, rhs, bin_operator
        end
    end

    def expr_statement
        expression simple_expr
    end

    def statement
        expr_statement
    end
end

f = File.open("test.nl", "r") 
l = Lexer.new f.read
p = Parser.new(l)
p.parse.each do |func_list|
    func_list.codegen
end
f.close